README
Queue-DBI

A queueing module with an emphasis on safety, using DBI as a storage system
for queued data.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Queue::DBI

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Queue-DBI

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Queue-DBI

    CPAN Ratings
        http://cpanratings.perl.org/d/Queue-DBI

    Search CPAN
        http://search.cpan.org/dist/Queue-DBI/


LICENSE AND COPYRIGHT

Copyright (C) 2009-2012 Guillaume Aubert

This program is free software; you can redistribute it and/or modify it
under the terms of the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Queue-DBI

v2.6.2  2015-03-28
        - Explicitly disallowed enqueuing scalars (GH-3).
        - Improved synopsis and documentation (GH-3).

v2.6.1  2014-11-15
        - Only use JSON::MaybeXS 1.001000 or above for testing (GH-2).

v2.6.0  2014-10-28
        - Added support for custom serializer methods, with Storable as the
          backward-compatible default (GH-1). Thank you Oliver Paukstadt!
        - Improved test coverage for Queue::DBI::Element->requeue_count().
        - Cleaned up the wording of error messages for incorrect arguments
          passed to Queue::DBI->new().

v2.5.3  2014-04-20
        - Fixed "package version matches dist version" for CPANTS.
        - Fixed license declared in meta files.
        - General cleanup.

v2.5.2  2013-05-13
        - Added missing prerequisite.

v2.5.1  2013-05-08
        - Moved release tests to xt/.
        - Updated Build.PL to handle release tests as part of disttest.
        - Added new release tests.
        - Formatted readme with markdown.
        - Added Test::FailWarnings to all tests.
        - Added info on how to contribute examples.
        - Updated links to point to MetaCPAN.
        - Updated links to point to GitHub's issue tracker.

v2.5.0  2012-10-31
        - (RT #79676) Fixed detection of SQL errors in selectrow_arrayref()
          calls.
        - Added Queue::DBI::Admin->assert_tables_verified().
        - (RT #79676) Queue::DBI::Admin now checks that the tables are
          correctly set up before performing operations.

v2.4.2  2012-10-10
        - Added Queue::DBI::Admin->has_table().
        - Added Queue::DBI::Admin->has_mandatory_fields().
        - (RT #79676) Added check that tables have the mandatory fields before
          deleting them, to prevent deletion of tables not related to
          Queue::DBI.

v2.4.1  2012-10-08
        - (RT #80055) Added PostgreSQL to the list of supported database types
          in the documentation.
        - (RT #80040) Added 'exclude_locked_elements' parameter to
          Queue::DBI->count().
        - Minor documentation cleanup.

v2.4.0  2012-10-06
        - Added full PostgreSQL support.
        - Refactored tests to allow testing against multiple database types.
        - Set up integration with Travis-CI for continuous integration.
        - Forced tests to be run recursively during build.

v2.3.1  2012-10-01
        - (RT #79676) Refactored Queue::DBI::Admin->has_tables() to check for
          mandatory fields and to simplify its return value.
        - (RT #79676) Force raising errors in try {} blocks.

v2.3.0  2012-09-25
        - (RT #79676) Added croak()s to propagate up all SQL errors.
        - (RT #79676) Documented return value of Queue::DBI->purge().
        - Added internal methods for Queue::DBI::Admin and tests.
        - Refactored code to leverage the new internal methods.
        - (RT #79676) Added Queue::DBI::Admin->drop_tables() and tests.
        - (RT #79676) Added Queue::DBI::Admin->has_tables() and tests.
        - POD cleanup.

v2.2.1  2012-09-19
        - Moved repository to GitHub.

v2.2.0  2012-09-18
        - (RT #78640) Fixed error message.
        - Removed non-standard quotes for identifiers when creating tables.
        - (RT #79676) Removed 'sqlite' parameter, create_tables() now detects
          automatically the database type.
        - (RT #79676) Fixed index names, to prevent conflicts.
        - (RT #79676) Added list of supported database types in documentation.
        - Drop tables in reverse creation order in create_tables(), to prevent
          issues with constraints.
        - (RT #79676) Fixed foreign key constraint when creating elements table
          to refer to the actual queues table name.

v2.1.1  2012-09-15
        - (RT #78640) Fixed bug in Queue::DBI->cleanup().

v2.1.0  2012-09-08
        - Re-licensed as GPLv3, to allow inclusion in Fedora.

v2.0.2  2012-08-31
        - Added missing prerequisite.

v2.0.1  2012-08-31
        - Added missing prerequisite.

v2.0.0  2012-08-31
        - (RT #78985) Added Queue::DBI::Admin, an interface to manage queues
          themselves. Thanks Sergey for suggesting this great extension!

v1.8.2  2012-08-17
        - (RT #78640) Deprecated dual-purpose methods to change Queue::DBI
          properties, replaced them with separate get and set methods:
            - replace max_requeue_count() with {get,set}_max_requeue_count();
            - replace lifetime() with {get,set}_lifetime();
            - replace verbose() with {get,set}_verbose().
        - Renamed Queue::DBI::Element->queue() into get_queue().

v1.8.1  2012-08-12
        - (RT #78640) Added Queue::DBI::Element->is_over_lifetime().
        - (RT #78640) Changed 'lifetime' parameter in Queue::DBI->purge() to be
          more consistent with retrieve_batch() and 'max_requeue_count',
          thanks Sergey for noticing!
        - Fixed MySQL-specific reference when retrieving the inserted queue
          element ID.
        - Added Queue::DBI::Element->get_created_time().
        - Fixed lifetime() to return undef if 'lifetime' is not set.
        - Fixed max_requeue_count() to return undef when 'max_requeue_count' is
          not set.
        - Fixed incorrect table name used in a failure case of success().
        - Cleaned up build_requires and requires.

v1.8.0  2012-08-05
        - (RT #78640) Added a 'lifetime' feature and a purge() function, thanks
          Sergey for suggesting these improvements.
        - Added purge() to clean queues.
        - Fixed how the last inserted ID is retrieved, for portability.
        - Added tests and code cleanup.
        - Refactored tests to use Test::Exception and subtests.

v1.7.3  2012-04-01
        - (RT #76194) Fixed typo in table name, thanks Sergey!

v1.7.2  2012-01-18
        - Cleanup to pass PerlCritic 'harsh'.
        - Fixed minor bug in error message.
        - Added acknowledgements.

v1.7.1  2011-12-16

t/LocalTest.pm
package LocalTest;

use strict;
use warnings;

use DBI;
use Test::More;


=head1 NAME

LocalTest - Test functions for L<Queue::DBI>.


=head1 VERSION

Version 2.6.2

=cut

our $VERSION = '2.6.2';


=head1 SYNOPSIS

	use lib 't/';
	use LocalTest;

	my $dbh = LocalTest::ok_database_handle();


=head1 FUNCTIONS

=head2 ok_database_handle()

Verify that a database handle can be created, and return it.

	my $dbh = LocalTest::ok_database_handle();

=cut

sub ok_database_handle
{
	$ENV{'QUEUE_DBI_DATABASE'} ||= 'dbi:SQLite:dbname=t/test_database||';

	my ( $database_dsn, $database_user, $database_password ) = split( /\|/, $ENV{'QUEUE_DBI_DATABASE'} );

	ok(
		my $database_handle = DBI->connect(
			$database_dsn,
			$database_user,
			$database_password,
			{
				RaiseError => 1,
			}
		),
		'Create connection to a database.',
	);

	my $database_type = $database_handle->{'Driver'}->{'Name'} || '';
	note( "Testing $database_type database." );

	return $database_handle;
}


=head1 AUTHOR

Guillaume Aubert, C<< <aubertg at cpan.org> >>.


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/Queue-DBI/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc LocalTest


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/Queue-DBI/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Queue-DBI>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Queue-DBI>

=item * MetaCPAN

L<https://metacpan.org/release/Queue-DBI>

=back


=head1 COPYRIGHT & LICENSE

Copyright 2009-2015 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'test' => {
              'TESTS' => 't/*.t t/01-Admin/*.t t/90-Element/*.t'
            },
  'PL_FILES' => {},
  'INSTALLDIRS' => 'site',
  'NAME' => 'Queue::DBI',
  'EXE_FILES' => [],
  'VERSION_FROM' => 'lib/Queue/DBI.pm',
  'PREREQ_PM' => {
                   'Try::Tiny' => 0,
                   'Test::Exception' => 0,
                   'Test::FailWarnings' => 0,
                   'DBD::SQLite' => 0,
                   'Data::Dumper' => 0,
                   'Storable' => 0,
                   'Carp' => 0,
                   'Test::More' => '0.94',
                   'MIME::Base64' => 0,
                   'Test::Type' => 0,
                   'Data::Validate::Type' => 0,
                   'DBI' => 0
                 }
)
;


lib/Queue/DBI/Element.pm
package Queue::DBI::Element;

use warnings;
use strict;

use Data::Dumper;
use Carp;


=head1 NAME

Queue::DBI::Element - An object representing an element pulled from the queue.


=head1 VERSION

Version 2.6.2

=cut

our $VERSION = '2.6.2';


=head1 SYNOPSIS

Please refer to the documentation for Queue::DBI.

=head1 METHODS

=head2 new()

Create a new Queue::DBI::Element object.

	my $element = Queue::DBI::Element->new(
		'queue'         => $queue,
		'data'          => $data,
		'id'            => $id,
		'requeue_count' => $requeue_count,
		'created'       => $created,
	);

All parameters are mandatory and correspond respectively to the Queue::DBI
object used to pull the element's data, the data, the ID of the element
in the database and the number of times the element has been requeued before.

It is not recommended for direct use. You should be using the following to get
Queue::DBI::Element objects:

	my $queue = $queue->next();

=cut

sub new
{
	my ( $class, %args ) = @_;

	# Check parameters
	foreach my $arg ( qw( data id requeue_count created ) )
	{
		croak "Argument '$arg' is needed to create the Queue::DBI object"
			if !defined( $args{$arg} ) || ( $args{$arg} eq '' );
	}
	croak 'Pass a Queue::DBI object to create an Queue::DBI::Element object'
		unless defined( $args{'queue'} ) && $args{'queue'}->isa( 'Queue::DBI' );

	# Create the object
	my $self = bless(
		{
			'queue'         => $args{'queue'},
			'data'          => $args{'data'},
			'id'            => $args{'id'},
			'requeue_count' => $args{'requeue_count'},
			'created'       => $args{'created'},
		},
		$class
	);

	return $self;
}


=head2 lock()

Locks the element so that another process acting on the queue cannot get a hold
of it

	if ( $element->lock() )
	{
		print "Element successfully locked.\n";
	}
	else
	{
		print "The element has already been removed or locked.\n";
	}

=cut

sub lock ## no critic (Subroutines::ProhibitBuiltinHomonyms)
{
	my ( $self ) = @_;
	my $queue = $self->get_queue();
	my $verbose = $queue->get_verbose();
	my $dbh = $queue->get_dbh();
	carp "Entering lock()." if $verbose;

	my $rows = $dbh->do(
		sprintf(
			q|
				UPDATE %s
				SET lock_time = ?
				WHERE queue_element_id = ?
					AND lock_time IS NULL
			|,
			$dbh->quote_identifier( $queue->get_queue_elements_table_name() ),
		),
		{},
		time(),
		$self->id(),
	) || croak 'Cannot lock element: ' . $dbh->errstr;

	my $success = ( defined( $rows ) && ( $rows == 1 ) ) ? 1 : 0;
	carp "Element locked: " . ( $success ? 'success' : 'already locked or gone' ) . "." if $verbose;

	carp "Leaving lock()." if $verbose;
	return $success;
}


=head2 requeue()

In case the processing of an element has failed

	if ( $element->requeue() )
	{
		print "Element successfully requeued.\n";
	}
	else
	{
		print "The element has already been removed or been requeued.\n";
	}

=cut

sub requeue
{
	my ( $self ) = @_;
	my $queue = $self->get_queue();
	my $verbose = $queue->get_verbose();
	my $dbh = $queue->get_dbh();
	carp "Entering requeue()." if $verbose;

	my $rows = $dbh->do(
		sprintf(
			q|
				UPDATE %s
				SET
					lock_time = NULL,
					requeue_count = requeue_count + 1
				WHERE queue_element_id = ?
					AND lock_time IS NOT NULL
			|,
			$dbh->quote_identifier( $queue->get_queue_elements_table_name() ),
		),
		{},
		$self->id(),
	);

	# Since Queue::DBI does not enclose the SELECTing of a queue_element
	# to be requeued, and this actual requeueing, it is possible for the
	# element to be requeued by another process in-between. It may even
	# be requeued, relocked, and successfully removed in-between. In either
	# case, the number of rows affected would be 0, and do() would return
	# 0E0, perl's "0 but true" value. This is not an error. However, if
	# -1 or undef is returned, DBI.pm encountered some sort of error.
	if ( ! defined( $rows ) || $rows == -1 )
	{
		# Always carp the information, since it is an error that
		# most likely doesn't come from this module.
		my $error = $dbh->errstr();
		carp 'Cannot requeue element: ' . ( defined( $error ) ? $error : 'no error returned by DBI' );
		return 0;
	}

	my $requeued = ( $rows == 1 ) ? 1 : 0;
	carp "Element requeued: " . ( $requeued ? 'done' : 'already requeued or gone' ) . "." if $verbose;

	# Update the requeue_count on the object as well if the database update was
	# successful.
	$self->{'requeue_count'}++
		if $requeued;

	carp "Leaving requeue()." if $verbose;
	return $requeued;
}


=head2 success()

Removes the element from the queue after its processing has successfully been
completed.

	if ( $element->success() )
	{
		print "Element successfully removed from queue.\n";
	}
	else
	{
		print "The element has already been removed.\n";
	}

=cut

sub success
{
	my ( $self ) = @_;
	my $queue = $self->get_queue();
	my $verbose = $queue->get_verbose();
	my $dbh = $queue->get_dbh();
	carp "Entering success()." if $verbose;

	# Possible improvement:
	# Add $self->{'lock_time'} in lock() and insist that it matches that value
	# when trying to delete the element here.

	# First, we try to delete the LOCKED element.
	my $rows = $dbh->do(
		sprintf(
			q|
				DELETE
				FROM %s
				WHERE queue_element_id = ?
					AND lock_time IS NOT NULL
			|,
			$dbh->quote_identifier( $queue->get_queue_elements_table_name() ),
		),
		{},
		$self->id(),
	);

	if ( ! defined( $rows ) || $rows == -1 )
	{
		croak 'Cannot remove element: ' . $dbh->errstr();
	}

	my $success = 0;
	if ( $rows == 1 )
	{
		# A LOCKED element was found and deleted, this is a success.
		carp "Found a LOCKED element and deleted it. Element successfully processed." if $verbose;
		$success = 1;
	}
	else
	{
		# No LOCKED element found to delete, try to find an UNLOCKED one in case it
		# got requeued by a parallel process.
		my $deleted_rows = $dbh->do(
			sprintf(
				q|
					DELETE
					FROM %s
					WHERE queue_element_id = ?
				|,
				$dbh->quote_identifier( $queue->get_queue_elements_table_name() ),
			),
			{},
			$self->id(),
		);

		if ( ! defined( $deleted_rows ) || $deleted_rows == -1 )
		{
			croak 'Cannot remove element: ' . $dbh->errstr;
		}

		if ( $deleted_rows == 1 )
		{
			# An UNLOCKED element was found and deleted. It probably means that
			# another process is still working on that element as well (possibly
			# because this element's lock timed-out, got cleaned up and picked by
			# another process).
			# Always carp for those, technically we processed the element successfully
			# so deleting it is the correct step to take, but we still want to throw
			# some warning for the user.
			carp 'Another process is probably working on the same element, as it was found UNLOCKED when we deleted it. '
				. 'Check parallelization issues in your code!';
			$success = 1;
		}
		else
		{
			# No element found at all. It probably means that another process had been
			# working on that element, but completed successfully its run and deleted
			# it.
			carp 'Another process has probably worked on this element and already deleted it after completing its operations. '
				. 'Check parallelization issues in your code!' if $verbose;
			$success = 0;
		}
	}

	carp "Leaving success()." if $verbose;
	return $success;
}


=head2 data()

Returns the data initially queued.

	my $data = $element->data();

=cut

sub data
{
	my ( $self ) = @_;

	return $self->{'data'};
}


=head2 requeue_count()

Returns the number of times that the current element has been requeued.

	my $requeue_count = $element->requeue_count();

=cut

sub requeue_count
{
	my ( $self ) = @_;

	return $self->{'requeue_count'};
}


=head2 id()

Returns the ID of the current element

	my $id = $element->id();

=cut

sub id
{
	my ( $self ) = @_;

	return $self->{'id'};
}


=head2 get_created_time()

Returns the unixtime at which the element was originally created.

	my $created = $element->get_created_time();

=cut

sub get_created_time
{
	my ( $self ) = @_;

	return $self->{'created'};
}


=head2 is_over_lifetime()

Returns a boolean indicating whether the current element is over the lifetime
specified when instanciating the queue. This is especially helpful if you
retrieve a large batch of elements and do long processing operations on each
of them.

	my $is_over_lifetime = $element->is_over_lifetime();

=cut

sub is_over_lifetime
{
	my ( $self ) = @_;
	my $queue = $self->get_queue();
	my $lifetime = $queue->get_lifetime();

	# If the queue doesn't a lifetime, an element will never "expire".
	return 0 if !defined( $lifetime );

	# Check the time the element was created.
	my $created_time = $self->get_created_time();
	return time() - $created_time > $lifetime;
}


=head1 INTERNAL METHODS

=head2 get_queue()

Returns the Queue::DBI object used to pull the current element.

	my $queue = $element->get_queue();

=cut

sub get_queue
{
	my ( $self ) = @_;

	return $self->{'queue'};
}


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/Queue-DBI/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc Queue::DBI::Element


You can also look for information at:

=over 4

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/Queue-DBI/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Queue-DBI>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Queue-DBI>

=item * MetaCPAN

L<https://metacpan.org/release/Queue-DBI>

=back


=head1 AUTHOR

L<Guillaume Aubert|https://metacpan.org/author/AUBERTG>,
C<< <aubertg at cpan.org> >>.


=head1 ACKNOWLEDGEMENTS

I originally developed this project for ThinkGeek
(L<http://www.thinkgeek.com/>). Thanks for allowing me to open-source it!


=head1 COPYRIGHT & LICENSE

Copyright 2009-2015 Guillaume Aubert.

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/

=cut

1;


Build.PL
use 5.006;
use strict;
use warnings;

use Module::Build;


my $class = Module::Build->subclass(
	code =>
	q|
		# Override 'test' to support running extended tests in the xt/
		# directory.
		sub ACTION_test
		{
			my ( $self ) = @_;

			$self->recursive_test_files(1);

			$self->test_files( 't', 'xt' )
				if $ENV{'RELEASE_TESTING'};

			return $self->SUPER::ACTION_test();
		}

		# Force running extended tests when testing the distribution.
		sub ACTION_disttest
		{
			my ( $self ) = @_;

			local $ENV{ RELEASE_TESTING } = 1;

			return $self->SUPER::ACTION_disttest();
		}
	|,
);

my $builder = $class->new(
	module_name          => 'Queue::DBI',
	license              => 'GPL_3',
	dist_author          => q{Guillaume Aubert <aubertg@cpan.org>},
	dist_version_from    => 'lib/Queue/DBI.pm',
	build_requires       =>
	{
		'DBD::SQLite'          => 0,
		'DBI'                  => 0,
		'Test::Exception'      => 0,
		'Test::FailWarnings'   => 0,
		'Test::More'           => 0.94,
		'Test::Type'           => 0,
	},
	requires             =>
	{
		'perl'                 => 5.006,
		'Data::Dumper'         => 0,
		'Carp'                 => 0,
		'Storable'             => 0,
		'MIME::Base64'         => 0,
		'Data::Validate::Type' => 0,
		'Try::Tiny'            => 0,
	},
	add_to_cleanup       =>
	[
		'Queue-DBI-*',
	],
	create_makefile_pl   => 'traditional',
	configure_requires   =>
	{
		'Module::Build'        => 0,
	},
	meta_merge           =>
	{
		resources              =>
		{
			repository  => 'https://github.com/guillaumeaubert/Queue-DBI',
			homepage    => 'https://metacpan.org/release/Queue-DBI',
			bugtracker  => 'https://github.com/guillaumeaubert/Queue-DBI/issues',
		},
		no_index                =>
		{
			file        =>
			[
				't/LocalTest.pm',
			],
		},
	},
	recursive_test_files => 1,
);

$builder->create_build_script();
